[STORM-3588] add GenericResourceAwareSchedulingPriorityStrategy to accommodate generic resource in grading topologies

RuiLi8080 opened the pull request:
Add GenericResourceAwareSchedulingPriorityStrategy by extending DefaultSchedulingPriorityStrategy to accommodate generic resources when sorting topologies.

TestGenericResourceAwareSchedulingPriorityStrategy provides situations where DefaultSchedulingPriorityStrategy wrongfully pre-empted all other running topologies. This is prevented by GenericResourceAwareSchedulingPriorityStrategy.

RuiLi8080 commented:
Add evictedTopologies set in ResourceAwareScheduler currently only for test since we did not find a better way to check eviction history. We added here considering that eviction is a RAS specific method and this new field could probably be useful in the future.

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/DefaultSchedulingPriorityStrategy.java:
```diff
@@ -140,7 +140,7 @@ public int compare(SimulatedUser o1, SimulatedUser o2) {
      * First sort by Topology Priority, if there is a tie for topology priority, topology uptime is used to sort.
      */
-    private static class TopologyByPriorityAndSubmissionTimeComparator implements Comparator<TopologyDetails> {
+    protected static class TopologyByPriorityAndSubmissionTimeComparator implements Comparator<TopologyDetails> {
```
This can be `private` since it's not used elsewhere ?

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -0,0 +1,181 @@
+        public TopologyDetails simScheduleNextHighest() {
+            TopologyDetails td = tds.pop();
+            assignedCpu += td.getTotalRequestedCpu();
+            assignedMemory += td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap();
```
We could reply on `super.simScheduleNextHighest();` so everything in GrasSimulatedUser is gras specific, like what you did elsewhere

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -0,0 +1,181 @@
+        /**
+         * Get a score for the simulated user.  This is used to sort the users, by their highest priority topology.
+         * Only give user guarantees that will not exceed cluster capacity.
+         * Score of each resource type is calculated as: (Requested + Assigned - Guaranteed)/Available
```
Nice javadoc!

nit: `Available` -> `clusterAvailable`

Ethanlm commented on storm-server/src/test/java/org/apache/storm/scheduler/resource/TestUtilsForResourceAwareScheduler.java:
```diff
@@ -83,6 +88,10 @@ public void addSelfTo(Map<String, Map<String, Number>> fullPool) {
     }
 
+    public static TestUserResources userRes(String name, Map<String, Double> resources) {
+        return new TestUserResources(name, resources);
```
This reminds me that we don't have a unit test to test the case where users have the generic resource guarantee. But this can be a follow up work

RuiLi8080 replied:
Addressed all comments, thanks for the prompt review. As for this, yes we should have some tests on User class. Currently we only have a testResourcePoolUtilization. When I was coding I also extended the usage of pool utilization related functions to include generic resources as well. But after I realized that there is nowhere we are using it. I will filed another PR for removing them.

Ethanlm replied:
Right. But what I meant is unit tests for GenericResourceAwareSchedulingPriorityStrategy when users do have the generic resource guarantee

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -202,7 +210,7 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
                                 Collection<WorkerSlot> workersToEvict = workingState.getUsedSlotsByTopologyId(topologyEvict.getId());
                                 topologySchedulingResources.adjustResourcesForEvictedTopology(toSchedule, topologyEvict);
-
+                                evictedTopologies.add(topologyEvict.getId());
```
One last thing, we can have a boolean like `forTests`, which by default to false and we can set it to true from unit tests. This is to avoid unnecessary computation/operations. And make it clear that this is only for tests. 

As for `getEvictedTopologies`, we can throw exceptions if `forTests` is false.

RuiLi8080 replied:
Addressed.

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -248,8 +252,17 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
+        if (forTest) {
+            return this.evictedTopologies;
+        } else {
+            throw new Exception("Topology eviction check is only provided for test purposes");
```
This can be `UnsupportedOperationException`

RuiLi8080 replied:
Fixed. Thank you.

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -256,7 +256,8 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
             return this.evictedTopologies;
         } else {
-            throw new Exception("Topology eviction check is only provided for test purposes");
+            throw new UnsupportedOperationException(
```
`UnsupportedOperationException` is a runtime exception. `throws Exception` is not needed in the signature. So is in the unit tests like testDefaultSchedulingPriorityStrategyNotEvicting

RuiLi8080 replied:
Thanks for catching.

kishorvpatil commented on storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:
```diff
@@ -889,6 +889,20 @@ public double getClusterTotalMemoryResource() {
+            for (Map.Entry<String, Double> entry : supGenericResources.entrySet()) {
+                String resourceName = entry.getKey();
+                Double amount = entry.getValue();
+                ret.put(resourceName, ret.getOrDefault(resourceName, 0.0) + amount);
```
A stream API could make it simpler to represent aggregation of maps:.
`return supervisors.values().stream().flatMap(m -> m.getTotalGenericResources().entrySet().stream()).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Double::sum));`

RuiLi8080 replied:
Very good suggestions. Thank you and addressed.

kishorvpatil commented on storm-server/src/main/java/org/apache/storm/scheduler/ISchedulingState.java:
```diff
@@ -271,6 +273,13 @@ boolean wouldFit(
+    /**
+     * Get the total amount of generic resources (excluding CPU and memory) in cluster.
+     */
+    default Map<String, Double> getClusterTotalGenericResource() {
```
Method name should probably be `getClusterTotalGenericResources`  with s

kishorvpatil replied:
Cluster should have a private variable that avoids recalculating this often. 
Similarly, it should have Set<String> genericResourceTypes.
And method `getClusterGenericResourceTypes`

RuiLi8080 replied:
Addressed

kishorvpatil commented on storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:
```diff
@@ -889,6 +889,20 @@ public double getClusterTotalMemoryResource() {
     }
 
+    @Override
+    public Map<String, Double> getClusterTotalGenericResource() {
```
Method name should include s at the end.

RuiLi8080 replied:
Yes, addressed

kishorvpatil commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -0,0 +1,177 @@
+            super(other, cluster);
+
+            Map<String, Double> guaranteedGenericResources = new HashMap<>();
+            Map<String, Double> availGenericResources = cluster.getClusterTotalGenericResource();   // generic resources that are offered
```
Instead of using `getClusterTotalGenericResource` here, probably adding below `getClusterTotalGenericResourceType` to cluster would help avoid calculations..
It would looks something like:
```    @Override
    public Set<String> getClusterGenericResourceTypes() {
        Set<String> resourceTypes =
            supervisors.values().parallelStream().map(sup -> sup.getTotalGenericResources().keySet()).flatMap(Set::stream).collect(Collectors.toSet());
        return resourceTypes;
    }```

RuiLi8080 replied:
Added

RuiLi8080 commented:
@kishorvpatil Thanks for your review. As I addressed your comment, I noticed that we actually can cache all types of total cluster resources since the superviors snapshot won't change in cluster object.
So we don't need to add another getClusterGenericResourceTypes() function. We can just cache generic resources map and only fetch the keySet while we need it.
I also added cache for CPU and memory in latest commit, any further comments would be appreciated.

kishorvpatil commented on storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:
```diff
@@ -80,6 +80,13 @@
+     */
+    private final double totalCpuResource;
+    private final double totalMemoryResource;
+    private final Map<String, Double> totalGenericResources;
```
These three variables nee to be part if copying from src in copy constructor and  `public Cluster(Cluster src, Topologies topologies)`

RuiLi8080 replied:
Fixed

bipinprasad replied:
If these variables are being used to keep running totals, should these be called " cacheTotalCpuResource" and so on. And the totals be incrementally updated in assign and unassign? If not running cache, then maybe just getTotalResouceCpu() == computeClusterCpuResource()?

kishorvpatil commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -0,0 +1,178 @@
+                String resource = entry.getKey();
+                Double requestedAmount = entry.getValue();
+                assignedGenericResources.put(resource, assignedGenericResources.getOrDefault(resource, 0.0) + requestedAmount);
+            }
```
should we not adjust cpu and memory as well? Very similar to how `SimulatedUser.simScheduleNextHighest()`

```                assignedCpu += td.getTotalRequestedCpu();
                assignedMemory += td.getTotalRequestedMemOffHeap() + td.getTotalRequestedMemOnHeap();
```
Even better would be generally refactor this whole class to minimize the duplicate code...

RuiLi8080 replied:
cpu and mem are adjusted by its overrided method on line 105:
```TopologyDetails td = super.simScheduleNextHighest();```

Not sure if this is what you are looking for.

bipinprasad commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -0,0 +1,178 @@
+            this.cpuAvail = cpuAvail;
+            this.memAvail = memAvail;
+            this.genericAvail = genericAvail;
+        }
```
I think this can be removed. Equivalent to Comparator<GrasSimulatedUser> c = (o1,o2) -> Double.compare(o1.getScore(cpuAvail, memAvail, genericAvail), o2.getScore(cpuAvail, memAvail, genericAvail));

bipinprasad commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    private static class GrasSimulatedUserComparator implements Comparator<GrasSimulatedUser> {
```
Is this comparator supposed to compare GrasSimulatedUser for use in a sort? Then it should compare o1 to o2. But it "seems" to be comparing  "distance" from some fixed point to o1 and o2. Is this intentional?

RuiLi8080 replied:
It is comparing GrasSimulatedUser  for sorting at line 58. And it is comparing o1 and o2 based on their scores, please refer to the overriding compare method at line 175.

bipinprasad commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -56,6 +59,10 @@
+    // record evicted topologies on each scheduling round, only used in test purpose now
+    private boolean forTest = false;
+    private Set<String> evictedTopologies = new HashSet<>();
+
```
Can this be removed. Use TestUtilsForResourceAwareScheduler#assertTopologiesNotScheduled.

RuiLi8080 replied:
Unfortunately no, I have to add this to check whether topo was evicted since in some scenarios, evicted topo would be put back in the same round of scheduling.

kishorvpatil commented:
@RuiLi8080 
I would like us to have behavior what we test and test that check production behavior. We should expose any evictions as standard result of scheduling available for both production and tests instead of forTest kind of special code path executed only during tests.

RuiLi8080 commented:
@kishorvpatil I think your point makes sense. But I was worried that maintaining the information of evicted topos will add unnecessary computation and also flood logs with evicted topoids since it is calculated at every round of scheduling. I made a small change so we only do this when testing or debugging, not sure if this can address your point.

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -113,6 +119,9 @@ public void schedule(Topologies topologies, Cluster cluster) {
                 User submitter = userMap.get(td.getTopologySubmitter());
                 scheduleTopology(td, cluster, submitter, orderedTopologies);
+                if (!evictedTopologies.isEmpty()) {
+                    LOG.warn("Evicted Topologies {} when scheduling topology: {}", evictedTopologies, td.getId());
```
evictedTopologies is cumulated during scheduling all the topologies since it only clears `evictedTopologies` at Line 111. So `evictedTopologies` here include all the evictedTopologies not only evicted by scheduling this topology `td.getId()`

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -239,6 +246,10 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
         markFailedTopology(topologySubmitter, cluster, td, "Failed to schedule within " + maxSchedulingAttempts + " attempts");
     }
 
+    public Set<String> getEvictedTopologies() {
```
I am seeing a possible race condition here. Since the `evictedTopologies` will be kept updated during scheduling, any function call to `getEvictedTopologies` doesn't guarantee to return a useful result.  This `evictedTopologies` includes every topology that is evicted during the scheduling process until the point when `getEvictedTopologies` is called.  It doesn't seem useful. 

To know what topologies are being evicted, we already have
```
LOG.debug("Evicting Topology {} with workers: {} from user {}", topologyEvict.getName(), workersToEvict,
                                    topologyEvict.getTopologySubmitter());
```
code in place. We can change it to `warn` if we want it to show up in the regular logs.  

@kishorvpatil  what do you think?

bipinprasad replied:
In that case this method should return a copy "return new HashSet<>(this.evictedTopologies);"

Ethanlm replied:
Yes that's right. We should return a unmodifiableCollection to solve the race condition. I think my question is mostly on how do we want to use this `evictedTopologies` information in regular code except unit test?

bipinprasad replied:
As a member of SchedulingResult this information can be more useful.
But looks like the intent is to keep track of topologies evicted as part of each topo scheduled, in which case we need a map i.e. topoId -> toposEvicted

Ethanlm commented on storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/priority/TestGenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -210,4 +216,11 @@ private Cluster mkTestCluster(Topologies topologies, Config config) {
     }
+
+    private Set<String> collectMapValues(Map<String, Set<String>> map) {
+        return map.values()
```
This looks overly complicated. 
```
        Set<String> set = new HashSet<>();
        map.values().forEach((s) -> set.addAll(s));
        return set;
```

RuiLi8080 replied:
Fixed.

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -196,21 +204,18 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
-                                LOG.debug("Evicting Topology {} with workers: {} from user {}", topologyEvict.getName(), workersToEvict,
-                                    topologyEvict.getTopologySubmitter());
-                                evictedSomething = true;
+                                tmpEvictedTopos.add(topologyEvict.getId());
```
`tmpEvictedTopos` is a set of evictedTopologies in every attempt. 
It might evict multiple topologies in maxSchedulingAttempts. 

So the current evictedTopologiesMap is just a subset.

RuiLi8080 replied:
Thanks for catching. This is my mistake. Made changes that 
1. new a hashmap for tmpEvictedTopos  at every round of scheduling.
2. combine evicted topologies at each attempt into tmpEvictedTopologiesMap

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -115,6 +122,7 @@ public void schedule(Topologies topologies, Cluster cluster) {
                 scheduleTopology(td, cluster, submitter, orderedTopologies);
             }
         }
+        evictedTopologiesMap = tmpEvictedTopologiesMap;
```
The idea of making `evictedTopologiesMap` a result after a complete schedule round is a good idea. 

But the code here makes `evictedTopologiesMap` and `tmpEvictedTopologiesMap` the same object.  
So tmpEvictedTopologiesMap.clear == evictedTopologiesMap.clear. This won't achieve the goal here.

RuiLi8080 replied:
Fixed on above comments.

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -196,21 +204,18 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
-                                Collection<WorkerSlot> workersToEvict = workingState.getUsedSlotsByTopologyId(topologyEvict.getId());
                                 topologySchedulingResources.adjustResourcesForEvictedTopology(toSchedule, topologyEvict);
-
-                                LOG.debug("Evicting Topology {} with workers: {} from user {}", topologyEvict.getName(), workersToEvict,
```
We can change this to `WARN` and add the current topologyId since this also includes the workers being evicted.

Then delete Line228.

RuiLi8080 replied:
As for log info. I feel the worker and user information is not useful here since all workers will be evicted when evicting topologies. We are more concerned about for what topology the evictions happen. And we can easily know user info from topology info.

Ethanlm replied:
Makes sense

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -248,6 +255,11 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
                     + topologySchedulingResources.getRemainingRequiredResourcesMessage());
     }
 
+    public Map<String, Set<String>> getEvictedTopologiesMap() {
```
I would add a comment like `This is not stable. It's subject to change` since we don't really use it elsewhere besides unit test currently. It might change depending on how we want to use the evictedTopologies information. As @bipinprasad  said, we might want to put it into the scheduleResult in the future.

RuiLi8080 replied:
Added.

Ethanlm replied:
Good comment

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -57,6 +59,8 @@
     private int schedulingTimeoutSeconds;
     private ExecutorService backgroundScheduling;
+    private Map<String, Set<String>> evictedTopologiesMap = new HashMap<>();   // topoId : topoEvicted
+    private Map<String, Set<String>> tmpEvictedTopologiesMap = new HashMap<>();
```
This doesn't have to be a attribute since it's only used inside schedule method

RuiLi8080 replied:
I was thinking not to add more parameters to he private scheduleTopology function, but I think this makes sense and fixed.

Ethanlm commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java:
```diff
@@ -219,8 +224,14 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
+                            if (!tmpEvictedTopologiesMap.containsKey(topoId)) {
+                                tmpEvictedTopologiesMap.put(topoId, new HashSet<>());
+                            }
+                            tmpEvictedTopologiesMap.get(topoId).addAll(tmpEvictedTopos);
```
Can simplify this with 

```
tmpEvictedTopologiesMap.computeIfAbsent(topoId, k -> new HashSet<>()).addAll(tmpEvictedTopos);
```

RuiLi8080 replied:
Fixed

govind-menon commented on storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java:
```diff
@@ -0,0 +1,178 @@
+        }
+
+        protected double getScore(double availableCpu, double availableMemory, Map<String, Double> availableGenericResources) {
+            TopologyDetails td = getNextHighest();
```
Nit: could we inline this to avoid unnecessary object creation

RuiLi8080 replied:
Fixed the inline.

As for `tmpEvictedTopologiesMap`, my concern is that in the the `scheduleTopology()` method, we have a lot of return/break points since we have a pretty complex nested if else structure. Passing the `tmpEvictedTopologiesMap` can make sure the snapshot will be passed to the `evictedTopologiesMap` after this round scheduling finished for whatever reason. 

So we create a new `tmpEvictedTopologiesMap` object before each round of `scheduleTopology()` and passed it over afterward.

govind-menon replied:
Thank you - I figured that was why - just wanted to be sure I wasn't missing anything

Ethanlm commented:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project storm-server: You have 1 Checkstyle violation. -> [Help 1]
```
checkstyle violations.

RuiLi8080 commented on storm-server/src/main/java/org/apache/storm/scheduler/ISchedulingState.java:
```diff
@@ -271,6 +273,13 @@ boolean wouldFit(
+     * Get the total amount of generic resources (excluding CPU and memory) in cluster.
+     */
+    default Map<String, Double> getClusterTotalGenericResources() {
+        return Collections.emptyMap();
```
No we don't. Only cluster class implement this interface and cluster will have the implementation in this PR. Good catch.

RuiLi8080 commented on storm-server/src/main/java/org/apache/storm/scheduler/Cluster.java:
```diff
@@ -873,20 +897,38 @@ public NormalizedResourceOffer getNonBlacklistedClusterAvailableResources(Collec
+        return supervisors.values().stream()
+            .map(sup -> sup.getTotalGenericResources().entrySet())
+            .flatMap(Set::stream)
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Double::sum));
```
Sorry I didn't think of a way to further simplify, could you elaborate? `supervisors.values().stream()` will create Stream<SupervisorDetails>. I don't how `Stream.of` can help here.

govind-menon replied:
If you see the example in the link it's a slightly different way to add the key, value pairs of a Map but it's not an issue. You can keep this as is.

RuiLi8080 commented:
Only one server test failed in JDK11, JDK8 passed.
